clustered/server/src/main/java/org/ehcache/clustered/server/EhcacheActiveEntity.java

```diff
@@ -293,6 +346,52 @@ private EhcacheEntityResponse invokeServerStoreOperation(ServerStoreOpMessage me
     }
   }
 
+  private void invalidate(ClientDescriptor originatingClientDescriptor, String cacheId, long key) {
+    int invalidationId = invalidationIdGenerator.getAndIncrement();
+    Set<ClientDescriptor> clientsToInvalidate = Collections.newSetFromMap(new ConcurrentHashMap<ClientDescriptor, Boolean>());
+    clientsToInvalidate.addAll(storeClientMap.get(cacheId));
+    clientsToInvalidate.remove(originatingClientDescriptor);
+
+    InvalidationHolder invalidationHolder = null;
+    if (stores.get(cacheId).getStoreConfiguration().getConsistency() == Consistency.STRONG) {
+      invalidationHolder = new InvalidationHolder(originatingClientDescriptor, clientsToInvalidate, cacheId, key);
+      clientsWaitingForInvalidation.put(invalidationId, invalidationHolder);
+    }
+
+    LOGGER.debug("SERVER: requesting {} client(s) invalidation of hash {} in cache {} (ID {})", clientsToInvalidate.size(), key, cacheId, invalidationId);
+    for (ClientDescriptor clientDescriptorThatHasToInvalidate : clientsToInvalidate) {
+      LOGGER.debug("SERVER: asking client {}" + clientDescriptorThatHasToInvalidate + " to invalidate hash {} from cache {} (ID {})", clientDescriptorThatHasToInvalidate, key, cacheId, invalidationId);
+      try {
+        clientCommunicator.sendNoResponse(clientDescriptorThatHasToInvalidate, clientInvalidateHash(cacheId, key, invalidationId));
+      } catch (MessageCodecException mce) {
+        //TODO: what should be done here?
+        LOGGER.error("Codec error", mce);
+      }
+    }
+
+    if (invalidationHolder != null && clientsToInvalidate.isEmpty()) {
+      clientInvalidated(invalidationHolder.clientDescriptorWaitingForInvalidation, invalidationId);
+    }
+  }
+
+  private void clientInvalidated(ClientDescriptor clientDescriptor, int invalidationId) {
+    InvalidationHolder invalidationHolder = clientsWaitingForInvalidation.get(invalidationId);
+
+    if (stores.get(invalidationHolder.cacheId).getStoreConfiguration().getConsistency() == Consistency.STRONG) {
+      invalidationHolder.clientsHavingToInvalidate.remove(clientDescriptor);
+      if (invalidationHolder.clientsHavingToInvalidate.isEmpty()) {
```

Comment:
Each invalidation has a unique `invalidationId` that is used as key of the `clientsWaitingForInvalidation` map, so all invalidation requests are isolated from each other.

As long as the `invalidationHolder.clientsHavingToInvalidate` `Set` is thread-safe, I fail to see how this code could be racy: the `Set` is filled up upon construction and entries are being removed while invalidation acks are received. Once the last ack is received (i.e.: the `Set` is empty), the `Map` gets cleaned up.